packages/backend/src/core/FinalizeExitEventsCollector.ts

```diff
@@ -0,0 +1,244 @@
+  data: FinalizeExitAction
+  blockNumber: number
+  minedAt: Timestamp
+}
+
+export class FinalizeExitEventsCollector {
+  constructor(
+    private readonly ethereumClient: EthereumClient,
+    private readonly forcedTransactionsRepository: ForcedTransactionsRepository,
+    private readonly transactionStatusRepository: TransactionStatusRepository,
+    // used only to sync finalized backwards
+    private readonly syncStatusRepository: SyncStatusRepository,
+    // used only to sync finalized backwards
+    private readonly logger: Logger,
+    private readonly perpetualAddress: EthereumAddress
+  ) {}
+
+  async collect(
+    blockRange: BlockRange
+  ): Promise<{ added: number; updated: number; ignored: number }> {
+    const minedFinalizes = await this.getMinedFinalizes(blockRange)
+    const results = await Promise.all(
+      minedFinalizes.map(async (finalize, i, array) => {
+        let previousFinalizeMinedAt =
+          (await this.forcedTransactionsRepository.findLatestFinalize()) ??
+          Timestamp(0)
+
+        if (i > 0) {
+          previousFinalizeMinedAt = array[i - 1].minedAt
+        }
+
+        return await this.processFinalizes(finalize, previousFinalizeMinedAt)
+      })
+    )
+
+    return results.reduce(
+      (acc, result) => ({ ...acc, [result]: acc[result] + 1 }),
+      { added: 0, updated: 0, ignored: 0 }
+    )
+  }
+
+  private async processFinalizes(
+    { hash, data, minedAt, blockNumber }: MinedTransaction,
+    previousWithdrawMinedAt: Timestamp
+  ) {
+    const connectedExit =
+      await this.forcedTransactionsRepository.findByFinalizeHash(hash)
+
+    if (
+      connectedExit &&
+      getTransactionStatus(connectedExit) === 'finalize sent'
+    ) {
+      await this.transactionStatusRepository.updateIfWaitingToBeMined({
+        hash,
+        mined: {
+          blockNumber,
+          at: minedAt,
+        },
+      })
+      return 'updated'
+    }
+
+    if (connectedExit) {
+      // This should never happen
+      return 'ignored'
+    }
+
+    const disconnectedExits =
+      await this.forcedTransactionsRepository.getWithdrawalsForFinalize(
+        data.starkKey,
+        minedAt,
+        previousWithdrawMinedAt
+      )
+
+    if (disconnectedExits.length === 0) {
+      // Someone did a regular withdraw that wasn't for a forced exit
+      return 'ignored'
+    }
+
+    await Promise.all(
+      disconnectedExits.map((exit) =>
+        this.forcedTransactionsRepository.saveFinalize(
+          exit.hash,
+          hash,
+          null,
+          minedAt,
+          blockNumber
+        )
+      )
+    )
+
+    return 'added'
+  }
+
+  // used only to sync finalized backwards
+  // #region sync-backwards
+  private syncExecuted = false
+  async oneTimeSync() {
+    if (this.syncExecuted) {
+      return
+    }
+
+    const firstSyncedBlock = await this.syncStatusRepository.getLastSynced()
+    if (!firstSyncedBlock || firstSyncedBlock < 14878490) {
+      return
+    }
```

Comment:
Doesn't this mean that we would run the oneTimeSync every time the server restarts? We should remember that we already done it by saving this in the database.

Reply:
I thought about this, but imo we can just delete this part after it syncs. Thats the purpose of all those comments.